core/src/main/java/org/apache/brooklyn/core/typereg/BasicManagedBundle.java

```diff
@@ -0,0 +1,175 @@
+
+    private String symbolicName;
+    private String version;
+    private String url;
+    private transient File localFileWhenJustUploaded;
+
+    /** Creates an empty one, with an ID, expecting other fields will be populated. */
+    public BasicManagedBundle() {}
+
+    public BasicManagedBundle(String name, String version, String url) {
+        if (name == null && version == null) {
+            Preconditions.checkNotNull(url, "Either a URL or both name and version are required");
+        } else {
+            Preconditions.checkNotNull(name, "Either a URL or both name and version are required");
+            Preconditions.checkNotNull(version, "Either a URL or both name and version are required");
+        }
+
+        this.symbolicName = name;
+        this.version = version;
+        this.url = url;
+    }
+    
+    @Override
+    public boolean isNameResolved() {
+        return symbolicName != null && version != null;
+    }
+    
+    @Override
+    public String getSymbolicName() {
+        return symbolicName;
+    }
+
+    public void setSymbolicName(String symbolicName) {
+        this.symbolicName = symbolicName;
+    }
+    
+    @Override
+    public String getVersion() {
+        return version;
+    }
+
+    public void setVersion(String version) {
+        this.version = version;
+    }
+    
+    @Override
+    public String getUrl() {
+        return url;
+    }
+
+    public void setUrl(String url) {
+        this.url = url;
+    }
+
+    /** This is cached on the object when just uploaded, then deleted after it has been persisted. */
+    @Beta
+    public void setTempLocalFileWhenJustUploaded(File localFileWhenJustUploaded) {
+        this.localFileWhenJustUploaded = localFileWhenJustUploaded;
+    }
+    @Beta
+    public File getTempLocalFileWhenJustUploaded() {
+        return localFileWhenJustUploaded;
+    }
+    
+    @Override
+    public String getOsgiUniqueUrl() {
+        return "brooklyn:"+getId();
+    }
+    
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                .add("symbolicName", symbolicName)
+                .add("version", version)
+                .add("url", url)
+                .toString();
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(symbolicName, version, url);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null) return false;
+        if (getClass() != obj.getClass()) return false;
+        OsgiBundleWithUrl other = (OsgiBundleWithUrl) obj;
+        if (!Objects.equal(symbolicName, other.getSymbolicName())) return false;
+        if (!Objects.equal(version, other.getVersion())) return false;
+        if (!Objects.equal(url, other.getUrl())) return false;
+        return true;
+    }
+
+    // ---
+    
+    @Override
+    public String getDisplayName() {
+        return null;
```

Comment:
no, display name is null and caller should find a sensible fallback to display, that's the convention elsewhere i think